packages/xrpc-server/src/stream/frames.ts

```diff
@@ -0,0 +1,122 @@
+import * as cborx from 'cbor-x'
+import * as uint8arrays from 'uint8arrays'
+import {
+  frameHeader,
+  FrameHeader,
+  FrameType,
+  InfoFrameHeader,
+  DataFrameHeader,
+  ErrorFrameHeader,
+  infoFrameBody,
+  InfoFrameBody,
+  ErrorFrameBody,
+  errorFrameBody,
+} from './types'
+
+export abstract class Frame {
+  header: FrameHeader
+  body: unknown
+  get op(): FrameType {
+    return this.header.op
+  }
+  toBytes(): Uint8Array {
+    return uint8arrays.concat([
+      cborx.encode(this.header),
+      cborx.encode(this.body),
+    ])
+  }
+  static fromBytes(bytes: Uint8Array) {
+    let i = 0
+    let header: unknown
+    let body: unknown = kUnset
+    cborx.decodeMultiple(bytes, (item) => {
+      if (i === 0) {
+        header = item
+      } else if (i === 1) {
+        body = item
+      } else {
+        throw new Error('Too many CBOR data items in frame')
+      }
+      i++
+    })
+    const parsedHeader = frameHeader.safeParse(header)
+    if (!parsedHeader.success) {
+      throw new Error(`Invalid frame header: ${parsedHeader.error.message}`)
+    }
+    if (body === kUnset) {
+      throw new Error('Missing frame body')
+    }
+    const frameOp = parsedHeader.data.op
+    if (frameOp === FrameType.Data) {
+      return new DataFrame({
+        type: parsedHeader.data.t,
+        body,
+      })
+    } else if (frameOp === FrameType.Info) {
+      const parsedBody = infoFrameBody.safeParse(body)
+      if (!parsedBody.success) {
+        throw new Error(`Invalid info frame body: ${parsedBody.error.message}`)
+      }
+      return new InfoFrame({
+        body: parsedBody.data,
+      })
+    } else if (frameOp === FrameType.Error) {
+      const parsedBody = errorFrameBody.safeParse(body)
+      if (!parsedBody.success) {
+        throw new Error(`Invalid error frame body: ${parsedBody.error.message}`)
+      }
+      return new ErrorFrame({
+        body: parsedBody.data,
+      })
+    } else {
+      const exhaustiveCheck: never = frameOp
+      throw new Error(`Unknown frame op: ${exhaustiveCheck}`)
+    }
+  }
+}
+
+export class DataFrame extends Frame {
+  header: DataFrameHeader
+  constructor(opts: { type?: number; body: unknown }) {
```

Comment:
Also can we make the `body` type a _little_ bit more stringent with `Record<string, unknown>`?
Or are there cases where the body is not a record?

also total nit, but kinda want the constructor of this to have the `body` as the first arg

constructor(body: Record<string, unknown>, opts: { type: number })

Reply:
The original thought was that the wire protocol doesn't need to put any limits on what you can transfer with it, as long as it encodes as CBOR (strings, numbers, arrays, raw bytes).  Then at the lexicon layer maybe it gets whittled down to whatever you can actually spec for a message, possibly just a record.  One way to split the difference here might be to make this generic with a default parameter of `Record<string, unknown>`.  I could see the generic here being pretty useful anyway:
```ts
export class DataFrame<T = Record<string, unknown>> extends Frame {
  header: DataFrameHeader
  constructor(opts: { type?: number; body: T }) {
```

Reply:
Hmmm yeah I take that point. I don't have a strong opinion. Down to leave it as is for simplicities sake 👌